rfd/0003-deduplication.md

```diff
@@ -0,0 +1,214 @@
+    boefje: Boefje
+    input_ooi: str | None = None
+    arguments: dict = {}
+    organization: str
+    runnable_hash: str | None = None
+    environment: dict[str, str] | None = None
+```
+
+The local job handler currently get the `BoefjeMeta` object from the scheduler
+and additionally sets `arguments["input"]`, `runnable_hash`, `environment`,
+`started_at` fields (see `BoefjeHandler` in boefjes/job_handler.py). The local
+boefje `run` function gets the full `BoefjeMeta` pydyantic object:
+
+```python
+from boefjes.job_models import BoefjeMeta
+
+def run(boefje_meta: BoefjeMeta) -> list[tuple[set, bytes | str]]:
+```
+
+The docker runner create the boefje meta object itself and currently only sets
+the `id`, `boefje`, `input_ooi`, `arguments`, `organization` and `environment` fields (see
+`create_boefje_meta` in boefjes/api.py). The docker boefje `run` function gets a
+dictionary with those values:
+
+```python
+def run(boefje_meta: dict) -> list[tuple[set, bytes | str]]:
+```
+
+### Arguments that are passed to boefjes that need to be considered for dedupliation
+
+To make sure we only do deduplication on boefje tasks that really are the same
+we should be more strict to what we pass to the boefje as input. We've been
+providing more information than is necessary to the boefjes by given the whole
+BoefjeMeta as input to each boefje.
+
+#### Usage of arguments field
+
+This is used by almost every boefje and the main way to specify on what the
+boefje should run. This is the serialized ooi, the result of calling `serialize`
+on an OOI.
+
+It might be better to provide the serialized ooi as `input_ooi` argument to the
+boefje `run` to make it more clear what the contents of the field is, see below
+for the proposed signature.
+
+#### Usage of organization field
+
+The external DB boefje currently needs the organization to fetch the data from
+the external database. Obviously this means the boefje can't be deduplicated.
+
+To support this we should add a flag to boefjes that require the organization.
+If the flag is set, we provide organization to the boefje and can't do
+deduplication. If the flag isn't set, we don't provide the organization and can
+duplicate. This way we make sure the organization is never used when it is not
+supposed to be.
+
+#### Usage of boefje task id field
+
+There is only one boefje that uses the task id. This
+
+The log4j vulnerability is now over 3 years old.
+
+For the sake of simplicity we might do away with passing the boefje the task id.
+
+#### Usage of input_ooi field
+
+Only the leakix boefje used the input_ooi field. For the sake of simplicy we
+should probably not provide this to the boefje. It should be easy to refactor
+this so it used the serialized input OOI.
+
+#### Usage of environment field
+
+This is not used directly from the BoefjeMeta, but this
+
+#### Usage of started_at, endated_at, boefje and runnable_hash fields
+
+This isn't used by any boefje and can be safely removed.
```

Comment:
The runnable hash should not be made available to the boefje itself, however, they are used by the rest of OpenKAT / can be used by the rest.